### PR TITLE
Feature/package install changes

### PR DESCRIPTION
**Pull request by brontide**
_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

**By submitting this pull request, I confirm the following (please check boxes, eg [X]):**
- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**
- [X] 7
---

Per issue #864 this allows users to see the installation error.

I think we would actually serve the users better by installing packages all at once since that's what package managers are good at.  So if there is agreement I can continue to work that angle.

### CODE REVIEW

File: automated install/basic-install.sh [hunk 1]
```diff
@@ -668,9 +668,10 @@ install_dependent_packages() {
 	declare -a argArray1=("${!1}")
 
 	for i in "${argArray1[@]}"; do
-		echo -n ":::    Checking for $i..."
-		package_check_install "${i}" &> /dev/null
-		echo " installed!"
+		echo -n ":::    Installing $i..."
+		PKG_MGR_OUT=$(package_check_install "${i}" 2>&1 ) && echo "Installed!" || (
```
Comment (dschaper) [inline automated install/basic-install.sh]: My bash is bad, will this variable be transitory and only live for the duration of the loop? Or does it need to be `local` to the function?

Comment (brontide) [inline automated install/basic-install.sh]: Dunno, I'll see if it behaves in a local, better to keep thing non-global id possible

**Comment by brontide** [comment]
This was tested against my stable of containers.  Some of them stuttered on various packages.  Restarting networking worked 100% of the time to complete the install.

```
root@pihole-u1404:~# ./basic-install.sh 
:::
::: You are root.
::: Verifying free disk space...
:::
::: Checking apt-get for upgraded packages.... done!
:::
::: Your system is up to date! Continuing with Pi-hole installation...
:::    Installing apt-utils...Installed!
:::    Installing whiptail...Installed!
:::    Installing git...Installed!
:::    Installing dhcpcd5...Installed!
:::    Installing dnsutils...Installed!
:::    Installing bc...Installed!
:::    Installing dnsmasq...Installed!
:::    Installing lighttpd...PACKAGE INSTALL ERROR
Reading package lists...
Building dependency tree...
Reading state information...
The following packages were automatically installed and are no longer required:
  libfreetype6 os-prober
Use 'apt-get autoremove' to remove them.
The following extra packages will be installed:
  gamin libgamin0 libterm-readkey-perl libterm-readline-perl-perl spawn-fcgi
Suggested packages:
  rrdtool apache2-utils
The following NEW packages will be installed:
  gamin libgamin0 libterm-readkey-perl libterm-readline-perl-perl lighttpd
  spawn-fcgi
0 upgraded, 6 newly installed, 0 to remove and 0 not upgraded.
Need to get 378 kB of archives.
After this operation, 1622 kB of additional disk space will be used.
Err http://archive.ubuntu.com/ubuntu/ trusty/universe libgamin0 amd64 0.1.10-4.1ubuntu1
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ trusty/universe gamin amd64 0.1.10-4.1ubuntu1
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ trusty/main libterm-readkey-perl amd64 2.31-1
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ trusty/universe libterm-readline-perl-perl all 1.0303-1
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ trusty/universe lighttpd amd64 1.4.33-1+nmu2ubuntu2
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com/ubuntu/ trusty/universe spawn-fcgi amd64 1.6.3-1
  Could not resolve 'archive.ubuntu.com'
Unable to correct missing packages.
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/g/gamin/libgamin0_0.1.10-4.1ubuntu1_amd64.deb  Could not resolve 'archive.ubuntu.com'

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/g/gamin/gamin_0.1.10-4.1ubuntu1_amd64.deb  Could not resolve 'archive.ubuntu.com'

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/libt/libterm-readkey-perl/libterm-readkey-perl_2.31-1_amd64.deb  Could not resolve 'archive.ubuntu.com'

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/libt/libterm-readline-perl-perl/libterm-readline-perl-perl_1.0303-1_all.deb  Could not resolve 'archive.ubuntu.com'

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/l/lighttpd/lighttpd_1.4.33-1+nmu2ubuntu2_amd64.deb  Could not resolve 'archive.ubuntu.com'

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/s/spawn-fcgi/spawn-fcgi_1.6.3-1_amd64.deb  Could not resolve 'archive.ubuntu.com'

E: Aborting install.
::: Sometimes this can be a transitory error, check connectivity and retry
```

## Fully tested

- pihole-debian-sid: fully functional
- pihole-debian8 (sysvinit ): fully functional
- pihole-fedora24: fully functional
- pihole-u1404: fully functional ( **this is new** thanks to #861 )
- pihole-u1604: fully functional
- pihole-u1610: fully functional
- pihole-centos7: fully functional

## Not working ( but were not before )

- pihole-debian7: Waiting on pull #863 
- pihole-alpine34: Waiting on pull #858 
- pihole-centos6: Need to resubmit pull


I can do the full monty on the platform detection and get debian 7 and

**Comment by dschaper** [comment]
And how about that consideration of download only the packages, does that seem to do anything different, or just punt the same problem down to another spot in the code? It's like the `apt update` doesn't take affect if done in a shell/subshell?

**Comment by brontide** [comment]
It *might* push the problem further down, but at least we would have fewer places to work on the problem and it seems like it would be easy to identify ( sudden loss of name resolution ).  But funny enough on the RPM based systems I often have trouble with PHP kicking out the first time as well so it's not just the networking packages.

**Comment by dschaper** [comment]
What I've noticed is that if there is a request for a package that is no longer in the repository (a new revision has been released and the client is requesting an older one) then `apt` will 404 since it can't get the package, and all repo's are at the core is a webserver. So the `apt` error of Can't find host is a bad coding response, the host is there, it's just returning a 404. Update the index via `apt update` and you get the new revision and the server sends the package with a 200. 

But we update the index at the start of the script, but it doesn't take effect. Restart the installer immediately and it uses the updated index and grabs the right package. It's really really odd. Next time you get a failure try and see if you can check the version that is being requested, and go to the repo and notice that there's another package version in its place...

**Comment by dschaper** [comment]
Revisiting this issue, thinking of the Docker cachebuster method. Update takes the array of packages, and builds out a space delineated list. Then `apt update && apt install <list>` so that we know that the frame / shell we are in has the latest indexes live and working.

**Comment by dschaper** [comment]
Cherry picking a few commits.

**Comment by dschaper** [comment]
Closing, please see #913

**Pull request closed without merging**